### DISCUSSION TITLE
Does argo-workflows support multiple controllers with different argo version in same k8s cluster ?

### DISCUSSION THREAD

**Discussion by wshi5985** [argoproj/argo-workflows#5103] [Q&A] [answered]
It looks like argo-workflows support multiple controllers in a single k8s cluster. 
https://github.com/argoproj/argo-workflows/blob/master/docs/workflow-controller-configmap.yaml#L12
Can multiple controllers in the same cluster has different argo version ? 
Thanks!

**Comment by sarabala1979** [comment]
yes, you can have a different version of controllers in the same cluster. you should configure the instance-id for each controller for sharding workflows for those controllers.

**Reply by wshi5985** [reply]
we have an existing controller with argo version 2.9.5. we would like to install second controller with latest argo version v3.0.0-rc1. we noticed that there are changes on CRDs (from example, apiVersion changed from apiextensions.k8s.io/v1beta1 to apiextensions.k8s.io/v1), CRDs are global, and names are consistent. 
How can we have 2 set of CRDs for different version of argo controllers ? Thanks.

**Reply by wshi5985** [reply]
not only the apiVersion changed, compare to v2.9.5, v3.0.0-rc1 CRD definition added schema/served/storage for version in multiple CRDs, 
  versions:
  - name: v1alpha1
    schema:
      openAPIV3Schema:
        properties:
          apiVersion:
            type: string
          kind:
            type: string
          metadata:
            type: object
          spec:
            type: object
            x-kubernetes-preserve-unknown-fields: true
        required:
        - metadata
        - spec
        type: object
    served: true
    storage: true

and removed "additionalPrinterColumns" in spec for CRD `workflows.argoproj.io`

If we use CRD definition for v3.0.0-rc1, will it breaks anything for the first argo controller(v2.9.5) ? Thanks.

**Comment by terrytangyuan** [comment, marked as answer]
As @sarabala1979 mentioned above, you can configure different controllers via different instance IDs.

If you need to serve different CRDs, you should probably try running different controllers in separate namespaces and configure your CRDs with `scope: Namespaced`: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions

You might also want to check out how to write a conversion webhook to convert and serve multiple versions of the CRDs: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion
